File: Data Structures & Algorithms/clone-graph/submission-5.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

public class Solution {
    public Node CloneGraph(Node node) {
		if (node == null) return null;
		Dictionary<Node, Node> originalToClone = new Dictionary<Node, Node>();
		Node dfs(Node original) {
			if (originalToClone.ContainsKey(original)) return originalToClone[original];

			Node clone = new Node(original.val);
			originalToClone[original] = clone;
			
			foreach (Node neighbor in original.neighbors) clone.neighbors.Add(dfs(neighbor));
			return clone;
		}
		return dfs(node);
    }
}

File: Data Structures & Algorithms/valid-tree/submission-5.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        
        adj_list = [[] for _ in range(n)]
        for v1, v2 in edges:
            adj_list[v1].append(v2)
            adj_list[v2].append(v1)
        visited = set()
       
        def dfs(node, parent):
            if node in visited:
                return False
            
            visited.add(node)
            for neighbor in adj_list[node]:
                if parent == neighbor:
                    continue
                if not dfs(neighbor, node):
                    return False
            return True
        
        return dfs(0, None) and len(visited) == n